rmf_battery/include/rmf_battery/agv/BatterySystem.hpp

```diff
@@ -0,0 +1,155 @@
+  {
+    /// The vehicle is powered by a Lead-Acid battery
+    LeadAcid,
+    /// The vehicle is powered by a Lithium-Ion battery
+    LiIon,
+    /// The vehicle is powered by a Nickel-Metal-Hydride battery
+    NiMH
+  };
+
+  class BatteryProfile
+  {
+  public:
+    /// The parameters required to construct a generic battery model to map
+    /// state of charge of the battery to its voltage.
+    ///
+    /// \param[in] resistance
+    ///   The internal resistance of the battery in ohms
+    ///
+    /// \param[in] max_voltage
+    ///   The maximum voltage in Volts of the battery
+    ///
+    /// \param[in] exp_voltage
+    ///   The voltage of the battery in Volts at the end of the exponential 
+    ///   zone in its discharge profile
+    ///
+    /// \param[in] exp_capacity
+    ///   The capacity of the battery in Ah at the end of the exponential zone
+    ///   in its discharge profile
+    ///
+    /// \param[in] nominal_capacity
+    ///   The capacity of the battery in Ah at the end of the exponential zone
+    ///   in its discharge profile
+    BatteryProfile(
+      double resistance,
+      double max_voltage,
+      double exp_voltage,
+      double exp_capacity,
+      double nominal_capacity,
+      double discharge_current);
+
+    BatteryProfile& resistance(double resistance);
+    double resistance() const;
+
+    BatteryProfile& max_voltage(double max_voltage);
+    double max_voltage() const;
+
+    BatteryProfile& exp_voltage(double exp_voltage);
+    double exp_voltage() const;
+
+    BatteryProfile& exp_capacity(double exp_capacity);
+    double exp_capacity() const;
+
+    BatteryProfile& nominal_capacity(double nominal_capacity);
+    double nominal_capacity() const;
+
+    BatteryProfile& discharge_current(double discharge_current);
+    double discharge_current() const;
+
+    /// Returns true if the values are valid, i.e. greater than zero.
+    bool valid() const;
+
+    class Implementation;
+  private:
+    rmf_utils::impl_ptr<Implementation> _pimpl;
+  };
+
+  /// Constructor
+  ///
+  /// \param[in] nominal_voltage
+  ///   The nominal voltage of the battery in volts
+  ///
+  /// \param[in] capacity
+  ///   The nominal capacity of the battery in Ah
+  ///
+  /// \param[in] charging_current
+  ///   The rated current in A for charging the battery
+  ///
+  /// \param[in] type
+  ///   The chemisty type of the battery
+  ///
+  /// \param[in] profile
+  ///   The battery profile for this battery. This is only required for calling
+  ///   the get_voltage() method
+  BatterySystem(
+    double nominal_voltage,
+    double capacity,
+    double charging_current,
+    BatteryType type = BatteryType::LeadAcid,
+    rmf_utils::optional<BatteryProfile> profile = rmf_utils::nullopt);
+
+  /// Estimate the voltage of the battery given its state of charge. This function
+  /// will return a nullopt if the battery profile is not defined
+  rmf_utils::optional<double> estimate_voltage(const double soc) const;
```

Comment:
Instead of doing this, I might suggest having a (non-member) function

```
double estimate_voltage(const BatterySystem& system, const BatteryProfile& profile);
```

Then this class doesn't have to store a `BatteryProfile`, assuming this function is the only reason to store one.

Reply:
Thanks for the suggestion. I'm considering getting rid of the `BatteryProfile` class along with `estimate_voltage()` function entirely. These are no longer being used in any downstream packages. We simply use the `state of charge` to keep track of the state of the battery.

Reply:
Removed BatteryProfile https://github.com/osrf/rmf_core/pull/217/commits/eb4ac7312e01824cf9381783455af4f128b065dd